docker/configure_workers_and_start.py

```diff
@@ -362,19 +380,137 @@ def add_worker_roles_to_shared_config(
-        # For now, all stream writers need http replication ports
-        instance_map[worker_name] = {
-            "host": "localhost",
-            "port": worker_port,
-        }
+    Args:
+        existing_dict: Either an existing worker template or a fresh blank one.
+        to_be_merged_dict: The template from WORKERS_CONFIGS to be merged into
+            existing_dict.
+    Returns: The newly merged together dict values.
+    """
+    new_dict: Dict[str, Any] = {}
+    for i in to_be_merged_dict.keys():
+        if (i == "endpoint_patterns") or (i == "listener_resources"):
+            # merge the two lists, remove duplicates
+            new_dict[i] = list(set(existing_dict[i] + to_be_merged_dict[i]))
+        elif i == "shared_extra_conf":
+            # merge dictionary's, the worker name will be replaced below after counting
+            new_dict[i] = {**existing_dict[i], **to_be_merged_dict[i]}
+        elif i == "worker_extra_conf":
+            # There is only one worker type that has a 'worker_extra_conf' and it is
+            # the media_repo. Since duplicate worker types on the same worker don't
+            # work, this is fine.
+            new_dict[i] = existing_dict[i] + to_be_merged_dict[i]
+        else:
+            # Everything else should be identical, like "app", which only works because
+            # all apps are now generic_workers.
+            new_dict[i] = to_be_merged_dict[i]
+    return new_dict
+
+
+def insert_worker_name_for_worker_config(
+    existing_dict: Dict[str, Any], worker_name: str
+) -> Dict[str, Any]:
+    """Insert a given worker name into the worker's configuration dict.
+
+    Args:
+        existing_dict: The worker_config dict that is imported into shared_config.
+        worker_name: The name of the worker to insert.
+    Returns: Copy of the dict with newly inserted worker name
+    """
+    dict_to_edit = existing_dict.copy()
+    for k, v in dict_to_edit["shared_extra_conf"].items():
+        # Only proceed if it's a string as some values are boolean
+        if isinstance(dict_to_edit["shared_extra_conf"][k], str):
+            # This will be ignored if the text isn't the placeholder.
+            dict_to_edit["shared_extra_conf"][k] = v.replace(
+                "placeholder_name", worker_name
+            )
+    return dict_to_edit
+
+
+def is_sharding_allowed_for_worker_type(worker_type: str) -> bool:
+    """Helper to check to make sure worker types that cannot have multiples do not.
+
+    Args:
+        worker_type: The type of worker to check against.
+    Returns: True if allowed, False if not
+    """
+    if worker_type in [
+        "background_worker",
+        "account_data",
+        "presence",
+        "receipts",
+        "typing",
+        "to_device",
+    ]:
+        return False
+    else:
+        return True
+
+
+def increment_counter(worker_type_counter: Dict[str, int], worker_type: str) -> int:
```

Comment:
rather than making a function for this, I'd be tempted to do (where you need it):

```
worker_type_counter = defaultdict(lambda _: 0)
```
then you can just easily `worker_type_counter[worker_type] += 1`

Reply:
`defaultdict(int)` will do that for you!

Reply:
I like it. Certainly better than a whole extra function to accomplish the same thing 😆 
Fixed.